lib/tasks/multitenant/tenants.rake

```diff
@@ -68,6 +68,36 @@ namespace :multitenant do
       update_tenant_ids(proc { |object| object.account.tenant_id }, proc { account }, proc { tenant_id == nil }, args.to_hash.merge({ table_name: 'Alert' }))
     end
 
+    desc 'Check and remove orphaned objects (whose tenant is missing), pass "destroy" argument to delete'
+    task :cleanup_orphans, [:mode] => :environment do |_task, args|
+      destroy = args[:mode] == "destroy"
+
+      puts "Checking orphaned objects..."
+      puts "WARNING: the found orphan objects will be destroyed" if destroy
+
+      provider_account_ids = Account.where(provider: true).pluck(:id) + [Account.master.id]
+
+      ApplicationRecord.descendants.each do |model|
```

Comment:
We can filter the classes further by filtering out children STI classes. Because if we urn the query with the base STI class, running with the children will be redundant. SOmething along the lines of:
```rb
     def base_models
      all_models = ApplicationRecord.descendants.select(&:arel_table).reject(&:abstract_class?)
      all_models.select! { _1.attribute_names.include? "tenant_id"}
      # we only want base STI classes, not the children
      all_models.select! do |model|
        base_class = model.base_class
        # either current model is the base_class or we can't find a base class amongst the discovered models /which would be very weird/
        base_class == model || all_models.none? { |potential_parent| potential_parent == base_class }
      end
    end
```

Reply:
Good idea, thanks!